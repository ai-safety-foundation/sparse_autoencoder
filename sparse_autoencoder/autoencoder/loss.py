"""Loss function for the Sparse Autoencoder."""
from jaxtyping import Float
import torch
from torch import Tensor
from torch.nn.functional import mse_loss


def reconstruction_loss(
    input_activations: Float[Tensor, "*batch input_activations"],
    output_activations: Float[Tensor, "*batch input_activations"],
) -> Tensor:
    """Reconstruction Loss (MSE).

    MSE reconstruction loss is calculated as the mean squared error between each each input vector
    and it's corresponding decoded vector. The original paper found that models trained with some
    loss functions such as cross-entropy loss generally prefer to represent features
    polysemantically, whereas models trained with MSE may achieve the same loss for both
    polysemantic and monosemantic representations of true features.

    Examples:
    >>> input_activations = torch.tensor([[3.0, 4]])
    >>> output_activations = torch.tensor([[1.0, 5]])
    >>> reconstruction_loss(input_activations, output_activations)
    tensor(2.5000)

    Args:
        input_activations: Input activations.
        output_activations: Reconstructed activations.

    Returns:
        Mean Squared Error reconstruction loss.
    """
    return mse_loss(input_activations, output_activations, reduction="mean")


def l1_loss(learned_activations: Float[Tensor, "*batch learned_activations"]) -> Tensor:
    """L1 Loss on Learned Activations.

    L1 loss penalty is the absolute sum of the learned activations. The L1 penalty is this
    multiplied by the l1_coefficient (designed to encourage sparsity).

    Examples:
    >>> learned_activations = torch.tensor([[2.0, -3]])
    >>> l1_loss(learned_activations)
    tensor([5.])

    Args:
        learned_activations: Activations from the hidden layer.

    Returns:
        L1 loss on learned activations.
    """
    return torch.abs(learned_activations).sum(dim=-1)


def sae_training_loss(
    reconstruction_loss_mse: Tensor,
    l1_loss_learned_activations: Tensor,
    l1_coefficient: float,
) -> Tensor:
    """Loss Function for the Sparse Autoencoder.

    The original paper used L2 reconstruction loss, plus l1 loss on the hidden (learned)
    activations.

    https://transformer-circuits.pub/2023/monosemantic-features/index.html#setup-autoencoder-motivation

    Examples:
    >>> reconstruction_loss_mse = torch.tensor([2.5000])
    >>> l1_loss_learned_activations = torch.tensor([1.])
    >>> l1_coefficient = 0.5
    >>> sae_training_loss(reconstruction_loss_mse, l1_loss_learned_activations, l1_coefficient)
    tensor(3.)

    Args:
        reconstruction_loss_mse: MSE reconstruction loss.
        l1_loss_learned_activations: L1 loss on learned activations.
        l1_coefficient: L1 coefficient. The original paper experimented with L1 coefficients of
            [0.01, 0.008, 0.006, 0.004, 0.001]. They used 250 tokens per prompt, so as an
            approximate guide if you use e.g. 2x this number of tokens you might consider using 0.5x
            the l1 coefficient.

    Returns:
        Overall training loss.
    """
    total_loss = reconstruction_loss_mse + l1_loss_learned_activations * l1_coefficient
    return total_loss.sum()
