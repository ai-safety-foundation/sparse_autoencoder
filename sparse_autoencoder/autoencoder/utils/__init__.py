"""Sparse autoencoder model & components utils."""
