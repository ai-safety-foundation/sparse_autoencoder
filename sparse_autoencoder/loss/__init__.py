"""Loss Modules.

Loss modules are specialised PyTorch modules that calculate the loss for a Sparse Autoencoder. They
all inherit from AbstractLoss, which defines the interface for loss modules and some common methods.

If you want to create your own loss function, see :class:`AbstractLoss`.

For combining multiple loss modules into a single loss module, see :class:`LossReducer`.
"""
