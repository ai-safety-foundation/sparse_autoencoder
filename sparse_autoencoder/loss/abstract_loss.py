"""Abstract loss."""
from abc import ABC, abstractmethod
from typing import TypeAlias, final

from strenum import LowercaseStrEnum
import torch
from torch.nn import Module

from sparse_autoencoder.tensor_types import (
    InputOutputActivationBatch,
    ItemTensor,
    LearnedActivationBatch,
    TrainBatchStatistic,
)


class LossReductionType(LowercaseStrEnum):
    """Loss reduction type (across batch items)."""

    MEAN = "mean"
    """Mean loss across batch items."""

    SUM = "sum"
    """Sum the loss from all batch items."""


LossLogType: TypeAlias = dict[str, int | float | str]
"""Loss log dict."""


class AbstractLoss(Module, ABC):
    """Abstract loss interface.

    Interface for implementing batch itemwise loss functions.
    """

    _modules: dict[str, "AbstractLoss"]  # type: ignore[assignment] (narrowing)
    """Children loss modules."""

    @abstractmethod
    def log_name(self) -> str:
        """Log name.

        Returns:
            Name of the loss module for logging.
        """

    @abstractmethod
    def forward(
        self,
        source_activations: InputOutputActivationBatch,
        learned_activations: LearnedActivationBatch,
        decoded_activations: InputOutputActivationBatch,
    ) -> TrainBatchStatistic:
        """Batch itemwise loss.

        Args:
            source_activations: Source activations (input activations to the autoencoder from the
                source model).
            learned_activations: Learned activations (intermediate activations in the autoencoder).
            decoded_activations: Decoded activations.

        Returns:
            Loss per batch item.
        """

    @final
    def batch_scalar_loss(
        self,
        source_activations: InputOutputActivationBatch,
        learned_activations: LearnedActivationBatch,
        decoded_activations: InputOutputActivationBatch,
        reduction: LossReductionType = LossReductionType.MEAN,
    ) -> ItemTensor:
        """Batch scalar loss.

        Args:
            source_activations: Source activations (input activations to the autoencoder from the
                source model).
            learned_activations: Learned activations (intermediate activations in the autoencoder).
            decoded_activations: Decoded activations.
            reduction: Loss reduction type. Typically you would choose LossReductionType.MEAN to
                make the loss independent of the batch size.

        Returns:
            Loss for the batch.
        """
        itemwise_loss = self.forward(source_activations, learned_activations, decoded_activations)

        match reduction:
            case LossReductionType.MEAN:
                return itemwise_loss.mean().squeeze()
            case LossReductionType.SUM:
                return itemwise_loss.sum().squeeze()

    def batch_scalar_loss_with_log(
        self,
        source_activations: InputOutputActivationBatch,
        learned_activations: LearnedActivationBatch,
        decoded_activations: InputOutputActivationBatch,
        reduction: LossReductionType = LossReductionType.MEAN,
    ) -> tuple[ItemTensor, LossLogType]:
        """Batch scalar loss.

        Args:
            source_activations: Source activations (input activations to the autoencoder from the
                source model).
            learned_activations: Learned activations (intermediate activations in the autoencoder).
            decoded_activations: Decoded activations.
            reduction: Loss reduction type. Typically you would choose LossReductionType.MEAN to
                make the loss independent of the batch size.

        Returns:
            Tuple of the batch scalar loss and a dict of any properties to log.
        """
        children_loss_scalars: list[ItemTensor] = []
        metrics: LossLogType = {}

        # If the loss module has children (e.g. it is a reducer):
        if len(self._modules) > 0:
            for loss_module in self._modules.values():
                child_loss, child_metrics = loss_module.batch_scalar_loss_with_log(
                    source_activations,
                    learned_activations,
                    decoded_activations,
                    reduction=reduction,
                )
                children_loss_scalars.append(child_loss)
                metrics.update(child_metrics)

            # Get the total loss & metric
            current_module_loss = torch.stack(children_loss_scalars).sum()

        # Otherwise if it is a leaf loss module:
        else:
            current_module_loss = self.batch_scalar_loss(
                source_activations, learned_activations, decoded_activations, reduction
            )

        # Add in the current loss module's metric
        log_name = self.log_name()
        metrics[log_name] = current_module_loss.detach().cpu().item()

        return current_module_loss, metrics

    @final
    def __call__(
        self,
        source_activations: InputOutputActivationBatch,
        learned_activations: LearnedActivationBatch,
        decoded_activations: InputOutputActivationBatch,
        reduction: LossReductionType = LossReductionType.MEAN,
    ) -> tuple[ItemTensor, LossLogType]:
        """Batch scalar loss.

        Args:
            source_activations: Source activations (input activations to the autoencoder from the
                source model).
            learned_activations: Learned activations (intermediate activations in the autoencoder).
            decoded_activations: Decoded activations.
            reduction: Loss reduction type. Typically you would choose LossReductionType.MEAN to
                make the loss independent of the batch size.

        Returns:
            Tuple of the batch scalar loss and a dict of any properties to log.
        """
        return self.batch_scalar_loss_with_log(
            source_activations, learned_activations, decoded_activations, reduction
        )
