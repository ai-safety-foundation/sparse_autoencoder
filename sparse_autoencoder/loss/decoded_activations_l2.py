"""L2 Reconstruction loss."""
from typing import final

from torch.nn.functional import mse_loss

from sparse_autoencoder.loss.abstract_loss import AbstractLoss
from sparse_autoencoder.tensor_types import (
    InputOutputActivationBatch,
    LearnedActivationBatch,
    TrainBatchStatistic,
)


@final
class L2ReconstructionLoss(AbstractLoss):
    """L2 Reconstruction loss.

    L2 reconstruction loss is calculated as the sum squared error between each each input vector
    and it's corresponding decoded vector. The original paper found that models trained with some
    loss functions such as cross-entropy loss generally prefer to represent features
    polysemantically, whereas models trained with L2 may achieve the same loss for both
    polysemantic and monosemantic representations of true features.

    Example:
        >>> import torch
        >>> loss = L2ReconstructionLoss()
        >>> input_activations = torch.tensor([[5.0, 4], [3.0, 4]])
        >>> output_activations = torch.tensor([[1.0, 5], [1.0, 5]])
        >>> unused_activations = torch.zeros_like(input_activations)
        >>> # Outputs both loss and metrics to log
        >>> loss(input_activations, unused_activations, output_activations)
        (tensor(11.), {'L2ReconstructionLoss': 11.0})
    """

    def forward(
        self,
        source_activations: InputOutputActivationBatch,
        learned_activations: LearnedActivationBatch,  # noqa: ARG002
        decoded_activations: InputOutputActivationBatch,
    ) -> TrainBatchStatistic:
        """Calculate the L2 reconstruction loss.

        Args:
            source_activations: Source activations (input activations to the autoencoder from the
                source model).
            learned_activations: Learned activations (intermediate activations in the autoencoder).
            decoded_activations: Decoded activations.

        Returns:
            Loss per batch item.
        """
        square_error_loss = mse_loss(source_activations, decoded_activations, reduction="none")

        # Sum over just the features dimension (i.e. batch itemwise loss). Note this is sum rather
        # than mean to be consistent with L1 loss (and thus make the l1 coefficient stable to number
        # of features).
        return square_error_loss.sum(dim=-1)
