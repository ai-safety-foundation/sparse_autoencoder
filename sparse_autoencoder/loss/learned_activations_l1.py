"""Learned activations L1 (absolute error) loss."""
from typing import final

import torch

from sparse_autoencoder.loss.abstract_loss import AbstractLoss
from sparse_autoencoder.tensor_types import (
    InputOutputActivationBatch,
    LearnedActivationBatch,
    TrainBatchStatistic,
)


@final
class LearnedActivationsL1Loss(AbstractLoss):
    """Learned activations L1 (absolute error) loss.

    L1 loss penalty is the absolute sum of the learned activations. The L1 penalty is this
    multiplied by the l1_coefficient (designed to encourage sparsity).

    Example:
        >>> l1_loss = LearnedActivationsL1Loss(0.1)
        >>> learned_activations = torch.tensor([[2.0, -3], [2.0, -3]])
        >>> unused_activations = torch.zeros_like(learned_activations)
        >>> # Returns loss and metrics to log
        >>> l1_loss(unused_activations, learned_activations, unused_activations)
        (tensor(0.5000), {'LearnedActivationsL1Loss': 0.5})
    """

    l1_coefficient: float
    """L1 coefficient."""

    def __init__(self, l1_coefficient: float) -> None:
        """Initialize the absolute error loss.

        Args:
            l1_coefficient: L1 coefficient. The original paper experimented with L1 coefficients of
                [0.01, 0.008, 0.006, 0.004, 0.001]. They used 250 tokens per prompt, so as an
                approximate guide if you use e.g. 2x this number of tokens you might consider using
                0.5x the l1 coefficient.
        """
        self.l1_coefficient = l1_coefficient
        super().__init__()

    def forward(
        self,
        source_activations: InputOutputActivationBatch,  # noqa: ARG002
        learned_activations: LearnedActivationBatch,
        decoded_activations: InputOutputActivationBatch,  # noqa: ARG002
    ) -> TrainBatchStatistic:
        """Learned activations L1 (absolute error) loss.

        Args:
            source_activations: Source activations (input activations to the autoencoder from the
                source model).
            learned_activations: Learned activations (intermediate activations in the autoencoder).
            decoded_activations: Decoded activations.

        Returns:
            Loss per batch item.
        """
        absolute_loss = torch.abs(learned_activations)

        return absolute_loss.sum(dim=-1) * self.l1_coefficient

    def extra_repr(self) -> str:
        """Extra representation string."""
        return f"l1_coefficient={self.l1_coefficient}"
