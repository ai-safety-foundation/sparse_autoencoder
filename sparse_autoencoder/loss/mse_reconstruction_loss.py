"""MSE Reconstruction loss."""
from typing import final

from torch.nn.functional import mse_loss

from sparse_autoencoder.loss.abstract_loss import AbstractLoss
from sparse_autoencoder.tensor_types import (
    InputOutputActivationBatch,
    LearnedActivationBatch,
    TrainBatchStatistic,
)


@final
class MSEReconstructionLoss(AbstractLoss):
    """MSE Reconstruction loss.

    MSE reconstruction loss is calculated as the mean squared error between each each input vector
    and it's corresponding decoded vector. The original paper found that models trained with some
    loss functions such as cross-entropy loss generally prefer to represent features
    polysemantically, whereas models trained with MSE may achieve the same loss for both
    polysemantic and monosemantic representations of true features.

    Example:
        >>> import torch
        >>> loss = MSEReconstructionLoss()
        >>> input_activations = torch.tensor([[5.0, 4], [3.0, 4]])
        >>> output_activations = torch.tensor([[1.0, 5], [1.0, 5]])
        >>> unused_activations = torch.zeros_like(input_activations)
        >>> # Outputs both loss and metrics to log
        >>> loss(input_activations, unused_activations, output_activations)
        (tensor(5.5000), {'MSEReconstructionLoss': 5.5})
    """

    def forward(
        self,
        source_activations: InputOutputActivationBatch,
        learned_activations: LearnedActivationBatch,  # noqa: ARG002
        decoded_activations: InputOutputActivationBatch,
    ) -> TrainBatchStatistic:
        """MSE Reconstruction loss (mean across features dimension).

        Args:
            source_activations: Source activations (input activations to the autoencoder from the
                source model).
            learned_activations: Learned activations (intermediate activations in the autoencoder).
            decoded_activations: Decoded activations.

        Returns:
            Loss per batch item.
        """
        square_error_loss = mse_loss(source_activations, decoded_activations, reduction="none")

        # Mean over just the features dimension (i.e. batch itemwise loss)
        return square_error_loss.mean(dim=-1)
