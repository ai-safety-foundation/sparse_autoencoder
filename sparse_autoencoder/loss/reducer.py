"""Loss reducer."""
from collections.abc import Iterator
from typing import final

from jaxtyping import Float
import torch
from torch import Tensor

from sparse_autoencoder.loss.abstract_loss import AbstractLoss
from sparse_autoencoder.tensor_types import (
    InputOutputActivationBatch,
    ItemTensor,
    LearnedActivationBatch,
)


@final
class LossReducer(AbstractLoss):
    """Loss reducer.

    Reduces multiple loss algorithms into a single loss algorithm (by summing). Analogous to
    nn.Sequential.

    Example:
        >>> from sparse_autoencoder.loss.decoded_activations_l2 import L2ReconstructionLoss
        >>> from sparse_autoencoder.loss.learned_activations_l1 import LearnedActivationsL1Loss
        >>> LossReducer(
        ...     L2ReconstructionLoss(),
        ...     LearnedActivationsL1Loss(0.001),
        ... )
        LossReducer(
          (0): L2ReconstructionLoss()
          (1): LearnedActivationsL1Loss(l1_coefficient=0.001)
        )

    """

    _modules: dict[str, "AbstractLoss"]
    """Children loss modules."""

    def log_name(self) -> str:
        """Log name.

        Returns:
            Name of the loss module for logging.
        """
        return "total_loss"

    def __init__(
        self,
        *loss_modules: AbstractLoss,
    ):
        """Initialize the loss reducer.

        Args:
            *loss_modules: Loss modules to reduce.

        Raises:
            ValueError: If the loss reducer has no loss modules.
        """
        super().__init__()

        for idx, loss_module in enumerate(loss_modules):
            self._modules[str(idx)] = loss_module

        if len(self) == 0:
            error_message = "Loss reducer must have at least one loss module."
            raise ValueError(error_message)

    def forward(
        self,
        source_activations: InputOutputActivationBatch,
        learned_activations: LearnedActivationBatch,
        decoded_activations: InputOutputActivationBatch,
    ) -> ItemTensor:
        """Reduce loss.

        Args:
            source_activations: Source activations (input activations to the autoencoder from the
                source model).
            learned_activations: Learned activations (intermediate activations in the autoencoder).
            decoded_activations: Decoded activations.

        Returns:
            Mean loss across the batch, summed across the loss modules.
        """
        all_modules_loss: Float[Tensor, "module train_batch"] = torch.stack(
            [
                loss_module.forward(source_activations, learned_activations, decoded_activations)
                for loss_module in self._modules.values()
            ]
        )

        return all_modules_loss.sum(dim=0)

    def __dir__(self) -> list[str]:
        """Dir dunder method."""
        return list(self._modules.__dir__())

    def __getitem__(self, idx: int) -> AbstractLoss:
        """Get item dunder method."""
        return self._modules[str(idx)]

    def __iter__(self) -> Iterator[AbstractLoss]:
        """Iterator dunder method."""
        return iter(self._modules.values())

    def __len__(self) -> int:
        """Length dunder method."""
        return len(self._modules)
