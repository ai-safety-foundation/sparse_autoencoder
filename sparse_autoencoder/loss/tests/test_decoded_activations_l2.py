"""Tests for the L2ReconstructionLoss class."""
import pytest
from syrupy.session import SnapshotSession
import torch

from sparse_autoencoder.autoencoder.utils.tensor_shape import shape_with_optional_dimensions
from sparse_autoencoder.loss.decoded_activations_l2 import L2ReconstructionLoss


@pytest.fixture()
def l2_loss() -> L2ReconstructionLoss:
    """Fixture for L2ReconstructionLoss."""
    return L2ReconstructionLoss()


def test_l2_loss_forward(l2_loss: L2ReconstructionLoss) -> None:
    """Test the forward method of L2ReconstructionLoss."""
    input_activations = torch.tensor([[5.0, 4.0], [3.0, 4.0]])
    output_activations = torch.tensor([[1.0, 5.0], [1.0, 5.0]])
    learned_activations = torch.zeros_like(input_activations)

    expected_loss = torch.tensor([8.5, 2.5])
    calculated_loss = l2_loss.forward(input_activations, learned_activations, output_activations)

    assert torch.allclose(calculated_loss, expected_loss), "L2 loss calculation is incorrect."


def test_l2_loss_with_zero_input(l2_loss: L2ReconstructionLoss) -> None:
    """Test the L2 loss function with zero inputs."""
    input_activations = torch.zeros((2, 3))
    output_activations = torch.zeros_like(input_activations)
    learned_activations = torch.zeros_like(input_activations)

    expected_loss = torch.zeros(2)
    calculated_loss = l2_loss.forward(input_activations, learned_activations, output_activations)

    assert torch.all(
        calculated_loss == expected_loss
    ), "L2 loss should be zero for identical zero inputs."


@pytest.mark.parametrize(
    ("input_shape", "learned_activations_shape"),
    [
        ((2, 3), (2, 6)),
        ((2, 4, 3), (2, 4, 6)),
    ],
)
def test_l2_loss_with_varying_input_shapes(
    l2_loss: L2ReconstructionLoss,
    input_shape: tuple[int],
    learned_activations_shape: tuple[int],
    snapshot: SnapshotSession,
) -> None:
    """Test the L2 loss function with varying input shapes."""
    torch.manual_seed(0)
    input_activations = torch.rand(shape_with_optional_dimensions(*input_shape))
    output_activations = torch.rand(shape_with_optional_dimensions(*input_shape))
    learned_activations = torch.rand(shape_with_optional_dimensions(*learned_activations_shape))

    calculated_loss = l2_loss.forward(input_activations, learned_activations, output_activations)

    rounded_zd_loss = torch.round(calculated_loss * 1e5).to(dtype=torch.int).tolist()
    assert str(rounded_zd_loss) == snapshot, "L2 loss has changed from the snapshot."


def test_l2_same_no_components_vs_1_component(l2_loss: L2ReconstructionLoss) -> None:
    """Test the L2 loss is the same with no components vs 1."""
    torch.manual_seed(0)
    input_activations = torch.rand(10, 3)
    learned_activations = torch.rand(10, 5)
    output_activations = torch.rand(10, 3)

    calculated_loss = l2_loss.forward(input_activations, learned_activations, output_activations)
    calculated_loss_1_component = l2_loss.forward(
        input_activations.unsqueeze(1),
        learned_activations.unsqueeze(1),
        output_activations.unsqueeze(1),
    )

    assert torch.allclose(
        calculated_loss, calculated_loss_1_component.squeeze(1)
    ), "L2 loss should be the same with no components vs 1."
