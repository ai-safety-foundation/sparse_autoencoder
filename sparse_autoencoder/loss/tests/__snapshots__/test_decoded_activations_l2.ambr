# serializer version: 1
# name: test_l2_loss_with_varying_input_shapes[input_shape0-learned_activations_shape0]
  '[5043, 10200]'
# ---
# name: test_l2_loss_with_varying_input_shapes[input_shape1-learned_activations_shape1]
  '[[26640, 3069, 3791, 10252], [26593, 12966, 17873, 21149]]'
# ---
